### PR TITLE
Checksum check for TF Binaries

### LINKED ISSUE #160: Verify Downloads Checksum & Signature
Due to the increasing frequency of supply chain attacks, it would be nice if tfswitch verified the checksums and gpg signatures of terraform when it downloads it for the first time, as described here:  
https://www.hashicorp.com/security#template-page-security:~:text=Release%20Archive%20Checksum%20Verification

Here's some example bash code that may provide a useful starting point for doing the equivalent in Go:
https://gist.github.com/markmsmith/cda59d5f24a812bea66fb3dbd7612397

### PR DESCRIPTION
**Pull request by MatrixCrawler**
This fixes #160 and #290 

- implement checksum check for terraform binaries
- implement signature check on terraform binaries
- added .gitignore for test with windows systems
- added .github action for testing with multiple go scenarios

### CODE REVIEW

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:15]: Would it make sense to allow users to customize the value?
E.g. if or when Hashicorp get the key rotated so that users don't depend on the new release of `tfswitch`?

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:18]: Same point as for `pubKeyId` above: would it make sense to make it adjustable by users?
Additionally users may want to have their own mirror (or e.g. buildfarms w/o access to external internet) and even re-packed (built by themselves) binaries.

So it looks like it makes sense to allow verification to be 1) customizable and 2) switchable. What do you think?

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:30]: Same `log.Fatal()` vs `fmt.Println()` question: why different approaches?

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:63]:
```suggestion
		log.Fatal("[Error]: Could not open the signature file")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:69]: It probs makes sense to remove trailing dot for consistency with the rest of messages
```suggestion
		log.Fatal("[Error]: Could not open the Terraform binary archive for signature verification")
```

File: test-data/checksum-check-file
Comment (yermulnik) [inline test-data/checksum-check-file:7]: It's best to add proper EOF (linefeed) please
image

File: lib/checksum.go
Comment (MatrixCrawler) [reply lib/checksum.go:62]: I was actually wondering mysql why we don't use the log class overall.
log.Fatal will automatically end with an sys.exit(1) and stops the execution. At this point it makes no sense to continue with execution.
I tried to consolidate all fmt.Println to log.Println or log.Fatal

File: lib/download.go
Comment (MatrixCrawler) [reply lib/download.go:15]: I think in this case we should cover that with a new release.

File: lib/checksum.go
Comment (yermulnik) [reply lib/checksum.go]: > but maybe we should talk about how to print the output in general.

💯 
I personally would prefer `[ERROR] Brief sum up: err text`
Same for info, warnings, etc: `[INFO] ...`, `[WARN] ...`, so on

File: lib/checksum.go
Comment (yermulnik) [reply lib/checksum.go]: I might be mixing up, though I seem to think it was @warrensbox's preference.
I'm all for `log.Fatal()` for fatal errors and `log.Println()` for the rest.

File: lib/download.go
Comment (yermulnik) [reply lib/download.go:15]: Agree. My point is that we're better off not introducing features that may block users as this had happened e.g. when Hashi moved to another site to distribute releases — that block all of the `tfswitch` users.

File: lib/utils.go
Comment (yermulnik) [inline lib/utils.go:1]: Would it make sense to name it according to the file name?
```suggestion
package utils
```

File: lib/checksum.go
Comment (MatrixCrawler) [reply lib/checksum.go]: We could have a look into https://github.com/op/go-logging for a loglevel based logging

File: lib/utils.go
Comment (MatrixCrawler) [reply lib/utils.go:1]: Then we would move this file into a new package. i don't think this would be necessary at this moment.
You might be mixing something up here

File: lib/checksum.go
Comment (yermulnik) [reply lib/checksum.go]: Yeah, that looks neat.
Would you be up to look into this when you get a chance? 😺

File: lib/utils.go
Comment (yermulnik) [reply lib/utils.go:1]: Okay, I'm not very familiar with this bits in Go. Just thought `utils` would better match helper functions.

File: lib/checksum.go
Comment (MatrixCrawler) [reply lib/checksum.go]: I will but I will create a new issue for that because this will be a lot of refactoring.

Comment (yermulnik) [reply lib/checksum.go]: Yep-yep, that's my point. Appreciate this.

File: lib/download.go
Comment (MatrixCrawler) [reply lib/download.go:18]: I just had a look into the possible configurations and i see a lot of consolidation potential there.
For example: the toml file only supports bin and version param (see #274), and i don't even see a test case for that (the test-data/test_tfswitchtoml/.tfswitch.toml) is unreferenced throughout the project.
i just started fiddling around with refactoring the parameter parsing but this is definitely out of scope of this ticket.

Comment (yermulnik) [reply lib/download.go:18]: Let me re-iterate and I'll leave this to you guys to decide then =)
https://github.com/warrensbox/terraform-switcher/pull/334#discussion_r1543158173 — we already were through the issue where `tfswitch` came up short for all of the users trying to download `terraform` because of the tiny change on the Hashicorp site end and users were absolutely blocked because they had no option to override the default. This is solely point that I frankly would not want to be introduced again.
While making this option customizable by users **looks to be** a scope of a separate ticket in general, I reckon the addition of a new functionality (which is the scope of this current PR) should better be accompanied with an option for users to override default values that configure this new functionality's behavior.
Thanks.

Comment (MatrixCrawler) [reply lib/download.go]: @yermulnik i made the options for the public key configurable. Will you pleas have a look at the changes?

Comment (crablab) [reply lib/download.go]: That's for the extra context @yermulnik - I vaguely remember that now. 

I think there's definitely a bigger thing, as @MatrixCrawler says, around configuration options that we could look at in a separate item.

**Review by crablab — APPROVED** [review]

File: lib/download.go
Comment (yermulnik) [reply lib/download.go]: @MatrixCrawler Looking. Would you mind resolving merge conflict meanwhile?

File: .github/workflows/go.yml
Comment (yermulnik) [inline .github/workflows/go.yml:1]: This file seems redundant as of existence of https://github.com/warrensbox/terraform-switcher/blob/master/.github/workflows/build.yml, right? 🤔

File: CHANGELOG.md
Comment (yermulnik) [inline CHANGELOG.md:9]: Do we have any automation around this? 🤔

File: lib/checksum.go
Comment (yermulnik) [inline lib/checksum.go:20]: I think it makes sense to sync with https://github.com/warrensbox/terraform-switcher/pull/350 so that we don't get a mess again when #350 is merged.
Would it make sense to complete #350 first and then get back to this PR?

**Review by yermulnik — CHANGES_REQUESTED** [review]
Let's complete #350 first to avoid discrepancies with logging again. WDT?

File: lib/download.go
Comment (yermulnik) [reply lib/download.go]: I'be partially reviewed and I think we should not slap this and #350 together and prefer #350 to get merged first.

File: CHANGELOG.md
Comment (crablab) [reply CHANGELOG.md:9]: I don't think so - probably not something for this PR, right?

Comment (yermulnik) [reply CHANGELOG.md:9]: Looks like yes. Let's remove this bit for now.
Ww'll need to investigate options on changelog automation later.

File: lib/checksum.go
Comment (MatthewJohn) [inline lib/checksum.go:63]: Is there any value in these return values, if log.Fatal immediately quits?

File: lib/download.go
Comment (MatthewJohn) [inline lib/download.go:102]: If this function is downloading arbitrary files now, zipFile probably isn't too accurate.

File: lib/download.go
Comment (MatthewJohn) [inline lib/download.go:75]: I'm not sure how much it matters - but I'm quite aware that if we download the public key, hash file and the zip file and the verification fails - all of these files are left in the install directory. Is it worth tidying them up after a failed download.

Comment (MatthewJohn) [reply lib/download.go:75]: In fact, the returned zip file path, once extracted by `Install()` will remove the zip file, but the checksum file will remain.

Comment (MatrixCrawler) [reply lib/download.go:81]: I delete the files needed for the verification after the process now.

File: CHANGELOG.md
Comment (MatrixCrawler) [reply CHANGELOG.md:9]: Removed this entry

File: lib/checksum.go
Comment (MatrixCrawler) [reply lib/checksum.go:76]: Yeah i will have to have a look into this once we merged the logger stuff from #350

File: lib/download.go
Comment (MatrixCrawler) [reply lib/download.go]: Fixed

**Comment by MatrixCrawler** [comment]
@yermulnik #350 is merged now and i rebased this. Will you have a look again please?

File: .github/workflows/go.yml
Comment (MatrixCrawler) [reply .github/workflows/go.yml:1]: Jep.. this is redundant. I will remove it.

File: lib/checksum.go
Comment (yermulnik) [inline lib/checksum.go:73]:
```suggestion
		logger.Errorf("Could not read armored key ring: %v", err)
```
1. Would it worth to provide some hint on how a user can try and resolve this?
2. Would it work on all supported OS (especially on Windows?)?

File: lib/checksum.go
Comment (yermulnik) [inline lib/checksum.go:72]: Would `defer closeFileHandlers(fileHandlersToClose)` work to replace repeatable `closeFileHandlers(fileHandlersToClose)` calls in this function?

File: lib/checksum.go
Comment (yermulnik) [inline lib/checksum.go:80]:
```suggestion
		logger.Errorf("Could not check detached signature: %v", err)
```

File: lib/checksum.go
Comment (yermulnik) [inline lib/checksum.go:83]:
```suggestion
	logger.Info("Checksum file signature verification successful.")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:16]: What do you think of consolidating default values at https://github.com/warrensbox/terraform-switcher/blob/master/lib/defaults.go or nearby the https://github.com/warrensbox/terraform-switcher/blob/master/main.go#L40-L51 ?

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:31]:
```suggestion
		logger.Error("Could not download public PGP key file.")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:38]:
```suggestion
		logger.Error("Could not download zip file.")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:45]:
```suggestion
		logger.Error("Could not download hash file.")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:52]:
```suggestion
		logger.Error("Could not download hash signature file.")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:58]:
```suggestion
		logger.Errorf("Could not open public key %q: %v", pubKeyFilename, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:64]:
```suggestion
		logger.Errorf("Could not open hash signature file %q: %v", hashSigFilePath, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:70]:
```suggestion
		logger.Errorf("Could not open zip file %q: %v", zipFilePath, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:76]:
```suggestion
		logger.Errord("Could not open hash file %q: %v", hashFilePath, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:87]:
```suggestion
		return "", errors.New("Signature of checksum file could not be verified")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:92]:
```suggestion
		return "", errors.New("Checksums did not match")
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:101]:
```suggestion
	logger.Infof("Downloading to %q", filepath.Join(installLocation, "/", fileName))
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:105]:
```suggestion
		logger.Errorf("Error downloading %s: %v", url, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:119]:
```suggestion
		logger.Errorf("Error creating %q: %v", filePath, err)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:141]:
```suggestion
			logger.Errorf("Error fetching public key file from %s", PubKeyUri)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:146]:
```suggestion
			logger.Errorf("Error renaming public key file from %q to %q", pubKeyFile, targetFileName)
```

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:158]:
```suggestion
			logger.Error("Error deleting %q: %v", path, err)
```

File: lib/install.go
Comment (yermulnik) [inline lib/install.go:127]: `logger.Fatalf()` performs `is.Exit(1)`, doesn't it? 🤔 
```suggestion
```

File: lib/utils.go
Comment (yermulnik) [inline lib/utils.go:7]: nitpick for consistency
```suggestion
// FileExists checks if a file exists and is not a directory before we try using it to prevent further errors
```

File: main.go
Comment (yermulnik) [inline main.go:68]: Since these are a bit different from other opts I'd separate them visually + consistency with other opts' descriptions
```suggestion

	getopt.StringVarLong(&lib.PubKeyId, "public-key-id", 'k', "The ID of the public key to check the checksums against")
	getopt.StringVarLong(&lib.PubKeyPrefix, "public-key-prefix", 'x', "The prefix of the public key. i.e. \"hashicorp_\"")
	getopt.StringVarLong(&lib.PubKeyUri, "public-key-uri", 'y', "The URI to download the public key from")
```
Also ain't public key is to verify the **signature** rather than the **checksums**? 😕

**Review by yermulnik — COMMENTED** [review]
Among other comments, I made my best to align logging messages

File: lib/install.go
Comment (MatrixCrawler) [reply lib/install.go:127]: I had the same phenomen. It should exit but it doesn't. Hence my manual exit calls.
Maybe I am missing a flag on the logger config or something.
I will have another look. Otherwise I will open an issue with slog

File: lib/checksum.go
Comment (MatrixCrawler) [reply lib/checksum.go:72]: I actually don't have a clue 😂 
my go knowledge is not that deep. I am learning with this project 😁

Comment (MatrixCrawler) [reply lib/checksum.go:72]: You are right. Will fix that

File: lib/download.go
Comment (yermulnik) [inline lib/download.go:85]: We probably can do something like the below instead of adding `cleanup(filesToCleanup)` three times? (I'm not very familiar with Go and hence might be mixing up things)
```suggestion
    defer cleanup(filesToCleanup)

	verified := checkSignatureOfChecksums(publicKeyFile, hashFile, signatureFile)
	if !verified {
		return "", errors.New("Signature of checksum file could not be verified")
	}

	match := checkChecksumMatches(hashFilePath, targetFile)
	if !match {
		return "", errors.New("Checksums did not match")
	}

	return zipFilePath, err
```

**Review by yermulnik — APPROVED** [review]

File: lib/download.go
Comment (MatrixCrawler) [reply lib/download.go]: Will call defer cleanup which will automatically called on return

**Comment by MatrixCrawler** [comment]
Thank you for all your effor @yermulnik 😄

**Pull request merged**
